packages/commerce-sdk-react/src/hooks/ShopperCustomers/cache.ts

```diff
@@ -17,81 +19,231 @@ import {
 import {and, pathStartsWith} from '../utils'
 
 type Client = ApiClients['shopperCustomers']
+type Customer = ShopperCustomersTypes.Customer
+type CustomerProductList = ShopperCustomersTypes.CustomerProductList
+type CustomerProductListResult = ShopperCustomersTypes.CustomerProductListResult
 
 const noop = () => ({})
 
-/** Invalidates the customer endpoint, but not derivative endpoints. */
-const invalidateCustomer = (parameters: Tail<QueryKeys['getCustomer']>): CacheUpdate => ({
-    invalidate: [{queryKey: getCustomer.queryKey(parameters)}]
-})
+// ReactQuery type def for the updater function uses type `unknown` for the first param,
+// this makes it hard to define the `createUpdateFunction` below, so we created this type.
+type Updater<T> = (data: T | undefined) => T | undefined
+
+// This is a slight variation of above, but we know the data is not null.
+type DataProcessor<T> = (data: NonNullable<T>) => T
+
+/**
+ * Create an update handler for the cache matrix. This updater will implicitly clone and assert
+ * that the data being passed to is is not undefined.
+ *
+ * @param update
+ * @returns
+ */
+const createUpdateFunction =
+    <T>(update: DataProcessor<T>): Updater<T> =>
+    (data) =>
+        !data ? data : update(clone(data))
 
 export const cacheUpdateMatrix: CacheUpdateMatrix<Client> = {
     createCustomerAddress(customerId, {parameters}, response) {
         // getCustomerAddress uses `addressName` rather than `addressId`
         const newParams = {...parameters, addressName: response.addressId}
         return {
-            // TODO: Rather than invalidate, can we selectively update?
-            ...invalidateCustomer(newParams),
-            update: [{queryKey: getCustomerAddress.queryKey(newParams)}]
+            update: [
+                {
+                    queryKey: getCustomerAddress.queryKey(newParams)
+                },
+                {
+                    queryKey: getCustomer.queryKey(newParams),
+                    updater: createUpdateFunction((customer: Customer) => {
+                        // Push new address onto the end of addresses list.
+                        customer?.addresses?.push(response)
```

Comment:
```suggestion
                        customer.addresses = [...customer.addresses ?? [], response]
```
If `addresses` doesn't exist, we need to create it. Also, don't need the null check on `customer` now that you've created the helper.

(Same comment for all the places you use this pattern.)

Reply:
So interestingly at this point in time, the customer itself could potentially be undefined (not likely in the PWA, integration of this library). So for situations where data coming in was undefined, I wasn't going out of my way to create data. 

It's possible that the customer is defined, but addresses isn't. So we might want to account for that, by doing something like. 

```
if (!customer.addresses) {
    customer.addresses = [...customer.addresses ?? [], response]
}
```


Thoughts?

Reply:
I replaces the push pattern with the array creation one. I did however have to add some conditional checks to make sure that the customer was defined as it's not technically guaranteed to exist